execution_engine/src/core/engine_state/mod.rs

```diff
@@ -128,6 +134,41 @@ where
         genesis_config_hash: Blake2bHash,
         protocol_version: ProtocolVersion,
         ee_config: &ExecConfig,
+    ) -> Result<GenesisResult, Error> {
+        if let Some(state_root_hash) = ee_config.state_root_hash() {
+            let missing_descendants = self
+                .state
+                .missing_trie_keys(correlation_id, state_root_hash)
+                .map_err(|err| Error::Exec(err.into()))?;
+            if !missing_descendants.is_empty() {
+                return Err(Error::Genesis(GenesisError::MissingStateKeys {
+                    state_root_hash,
+                    missing_descendants,
+                }));
+            }
+            self.commit_genesis_with_state(
+                correlation_id,
+                genesis_config_hash,
+                protocol_version,
+                ee_config,
+                state_root_hash,
+            )
+        } else {
+            self.commit_fresh_genesis(
+                correlation_id,
+                genesis_config_hash,
+                protocol_version,
+                ee_config,
+            )
+        }
+    }
+
+    pub fn commit_fresh_genesis(
```

Comment:
This might be naive, but isn't `commit_fresh_genesis` the same as `commit_genesis_with_state`, only `state_root_hash` is set to `self.state.empty_root()`?

Reply:
Explicitly:

```rust
pub fn commit_fresh_genesis(
        &self,
        correlation_id: CorrelationId,
        genesis_config_hash: Blake2bHash,
        protocol_version: ProtocolVersion,
        ee_config: &ExecConfig,
    ) -> Result<GenesisResult, Error> {
        commit_genesis_with_state(
          &self,
          correlation_id,
          genesis_config_hash,
          protocol_version,
          ee_config,
          self.state.empty_root(),
      )
}
```

Now that I write that, you would have to insert the empty pointer block into the trie-store so the empty root would be valid.  We could do that with `StorageProvider` as-is, and it would mean a lot less code duplication.